### ISSUE TITLE
restartBrowserBetweenTests should restart test in beforeEach, not afterEach

### ISSUE THREAD

**Issue by LucasSloan** [angular/protractor#4224]
The flag restartBrowserBetweenTests does what it says on the hood, but it does so in afterEach, which means that the state of the browser can't be keyed off of the test that it will be running.  For example we can't inject the test name into the browser at start time because it's starting up while the previous test is active.

**Comment by sjelin** [comment]
Fair enough.  I suppose the [hack we use](https://github.com/angular/protractor/blob/master/lib/frameworks/setupAfterEach.js#L26-L29) to setup `afterEach` in jasmine/mocha would actually be more reliable with `beforeEach`.  `afterEach` felt more consistent with with the [old implementation](https://github.com/angular/protractor/blob/master/lib/runner.ts#L443-L444) based on event emitters, but I guess that's only superficially true.
